test/hooks.jl

```diff
@@ -31,8 +31,68 @@
 
         # Check that Meta.parse throws the JuliaSyntax.ParseError rather than
         # Meta.ParseError when Core integration is enabled.
-        @test_throws JuliaSyntax.ParseError Meta.parse("[x")
+        @test_throws JuliaSyntax.ParseError Meta.parse("[x)")
 
         JuliaSyntax.enable_in_core!(false)
     end
+
+    @testset "Expr(:incomplete)" begin
+        JuliaSyntax.enable_in_core!()
+
+        @test Meta.isexpr(Meta.parse("[x"), :incomplete)
+
+        for (str, tag) in [
+                ""             => :none
+                "\""           => :string
+                "\"\$foo"      => :string
+                "#="           => :comment
+                "'"            => :char
+                "'a"           => :char
+                "`"            => :cmd
+                "("            => :other
+                "["            => :other
+                "begin"        => :block
+                "quote"        => :block
+                "let"          => :block
+                "let;"         => :block
+                "for"          => :other
+                "for x=xs"     => :block
+                "function"     => :other
+                "function f()" => :block
+                "macro"        => :other
+                "macro f()"    => :block
+                "f() do"       => :other
+                "f() do x"     => :block
+                "module"       => :other
+                "module X"     => :block
+                "baremodule"   => :other
+                "baremodule X" => :block
+                "mutable struct"    => :other
+                "mutable struct X"  => :block
+                "struct"       => :other
+                "struct X"     => :block
+                "if"           => :other
+                "if x"         => :block
+                "while"        => :other
+                "while x"      => :block
+                "try"          => :block
+                "try x catch"  => :block  # ambiguous - could be :block or :other
```

Comment:
IMHO `:block` is correct here, since `try x catch end` is valid.

Reply:
I thought `:other` also fits because `try x catch exc body end` parses. So could either be a block or a symbol for the exception name.  So this seemed ambiguous to me?

Reply:
(BTW, I don't think the REPL does much with the distinction here. Though the REPL completion code is a bit hard to follow.)

Reply:
`:other` doesn't make sense to me since the presence of `exc` is irrelevant for the validity of the code, but I also don't really know how this value actually ends up being used.

Reply:
Seems fair, I'll just change the note next to this to explain the possibility of both syntaxes.  AFAIK this doesn't change things much either way with the REPL (and in any case as noted in the description, I think `Base.incomplete_tag` itself is quite inconsistent in any case).